pdns/zonemd.cc

```diff
@@ -65,26 +28,117 @@ void pdns::zonemdVerify(const DNSName& zone, ZoneParserTNG& zpt, bool& validatio
       std::string err = "Bad record content in record for '" + dnsResourceRecord.qname.toStringNoDot() + "|" + dnsResourceRecord.qtype.toString() + "': " + e.what();
       throw PDNSException(err);
     }
-    if (dnsResourceRecord.qtype == QType::SOA && dnsResourceRecord.qname == zone) {
-      soaRecordContent = std::dynamic_pointer_cast<SOARecordContent>(drc);
+
+    if (dnsResourceRecord.qname == d_zone) {
+      switch (dnsResourceRecord.qtype) {
+      case QType::SOA:
+        d_soaRecordContent = std::dynamic_pointer_cast<SOARecordContent>(drc);
```

Comment:
I would feel better if the return value of `std::dynamic_pointer_cast` was checked. We have had issues with this before with records of a class != `QClass:IN` that were not properly filtered upfront.